vars/kanikoExecute.groovy

```diff
@@ -0,0 +1,127 @@
+     */
+    'customTlsCertificateLinks',
+    /**
+     * Defines the location of the Dockerfile relative to the Jenkins workspace.
+     */
+    'dockerfile',
+    /**
+     * Defines the id of the file credentials in your Jenkins credentials store which contain the file `.docker/config.json`.
+     * You can find more details about the Docker credentials in the [Docker documentation](https://docs.docker.com/engine/reference/commandline/login/).
+     */
+    'dockerConfigJsonCredentialsId',
+    /** @see dockerExecute */
+    'dockerEnvVars',
+    /** @see dockerExecute */
+    'dockerOptions',
+    /** @see dockerExecute */
+    'dockerImage'
+]
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+/**
+ * Executes a [Kaniko](https://github.com/GoogleContainerTools/kaniko) build for creating a Docker container.
+ */
+@GenerateDocumentation
+void call(Map parameters = [:]) {
+    handlePipelineStepErrors(stepName: STEP_NAME, stepParameters: parameters) {
+
+        final script = checkScript(this, parameters) ?: this
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+            .loadStepDefaults()
+            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+            .mixinStageConfig(script.commonPipelineEnvironment, parameters.stageName?:env.STAGE_NAME, STEP_CONFIG_KEYS)
+            .mixin(parameters, PARAMETER_KEYS)
+            .use()
+
+        new Utils().pushToSWA([
+            step: STEP_NAME
+        ], config)
+
+        def buildOptions = new GStringTemplateEngine().createTemplate(config.containerBuildOptions).make([config: config, env: env]).toString()
+
+        if (!buildOptions.contains('--destination')) {
+            if (config.containerImageNameAndTag) {
+                buildOptions += " --destination ${config.containerImageNameAndTag}"
+            } else {
+                buildOptions += " --no-push"
+            }
+        }
+
+        dockerExecute(
+            script: script,
+            containerCommand: config.containerCommand,
+            containerShell: config.containerShell,
+            dockerEnvVars: config.dockerEnvVars,
+            dockerImage: config.dockerImage,
+            dockerOptions: config.dockerOptions
+        ) {
+            // prepare kaniko container for running with proper Docker config.json and custom certificates
+            sh """#!${config.containerShell}
+${config.containerPreparationCommand}
+${getCertificateUpdate(config.customTlsCertificateLinks)}
+"""
+
+            def uuid = UUID.randomUUID().toString()
+            if (config.dockerConfigJsonCredentialsId) {
+                // write proper config.json with credentials
+                withCredentials([file(credentialsId: config.dockerConfigJsonCredentialsId, variable: 'dockerConfigJson')]) {
```

Comment:
out of curiosity: what is the reason for reading a file which is provided via the Jenkins credentials framework?
Some context would be appreciated.

Reply:
It is the Docker config file which contains the registry credentials.
Kaniko uses `.docker/config.json` for accessing e.g. private registries

Reply:
I updated the docs to explain the credential thing a bit more ...

Reply:
>It is the Docker config file which contains the registry credentials.

I don't get it. We have a file (somewhere in the project sources, or somewhere on the build servers file system, or somewhere inside a docker image). This file contains credentials. Right?

Now we configure the file name via the Jenkins CredentialsPlugin. Right?

In case I got it, I do not understand what is going on. Why does configuring a file name for a file in the sources |on file system | in docker image  making anything more secure. The credentials are contained in the config file anyway. But maybe I'm on the wrong track here ...

What I would understand is: having the config file with some placeholders and injecting the secrets. But having the secrets plain somewhere in a file is a security issue.

Reply:
complete file is handled by Jenkins not only the file name ...

Reply:
... understood, I was on a misleading track ...